crates/hir-ty/src/next_solver/interner.rs

```diff
@@ -1345,10 +1354,28 @@ impl<'db> rustc_type_ir::Interner for DbInterner<'db> {
         self,
         def_id: Self::DefId,
     ) -> EarlyBinder<Self, impl IntoIterator<Item = (Self::Clause, Self::Span)>> {
+        fn is_self_or_assoc(def_id: GenericDefId, ty: Ty<'_>) -> bool {
+            match ty.kind() {
+                rustc_type_ir::TyKind::Param(param) => {
+                    param.index == 0 && param.id.parent() == def_id
```

Comment:
You don't need to check `param.id.parent` (and in fact, I'd prefer if you didn't, because I want `id` to eventually go away). The index fully determines the parameter.